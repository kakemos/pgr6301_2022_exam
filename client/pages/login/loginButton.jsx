import { randomString } from "../../lib/randomString";
import { sha256 } from "../../lib/sha256";
import React from "react";

export function LoginButton({ config, label, provider }) {
  async function handleLogin() {
    const {
      authorization_endpoint,
      response_type,
      scope,
      client_id,
      code_challenge_method,
    } = config[provider];

    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);
    const redirect_uri = `${window.location.origin}/login/${provider}/callback`;

    const parameters = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri,
    };

    if (code_challenge_method) {
      const state = randomString(50);
      const code_verifier = randomString(50);
      window.sessionStorage.setItem("authorization_state", state);
      window.sessionStorage.setItem("code_verifier", code_verifier);
      parameters.code_challenge_method = code_challenge_method;
      parameters.code_challenge = await sha256(code_verifier);
    }

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(parameters);
  }

  return (
    <div>
      <button className={"login-button"} onClick={handleLogin}>
        {label}
      </button>
    </div>
  );
}
