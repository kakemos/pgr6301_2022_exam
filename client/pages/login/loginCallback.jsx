import { useNavigate, useParams } from "react-router-dom";
import React, { useContext, useEffect, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";

export function LoginCallback({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(NewsApiContext);

  useEffect(async () => {
    const { access_token, error, error_description, state, code } =
      Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1))
      );

    const expected_state = window.sessionStorage.getItem("expected_state");
    if (!state || expected_state !== state) {
      setError("Unexpected state");
      return;
    }

    if (error || error_description) {
      setError(`Error: ${error} (${error_description})`);
      return;
    }

    if (code) {
      const { client_id, token_endpoint } = config[provider];
      const code_verifier = window.sessionStorage.getItem("code_verifier");
      const state = window.sessionStorage.getItem("state");
      const redirect_uri = `${window.location.origin}/login/${provider}/callback`;
      const payload = {
        grant_type: "authorization_code",
        code,
        client_id,
        code_verifier,
        state,
        redirect_uri,
      };
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      if (!res.ok) {
        setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
        return;
      }
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });
      reload();
      navigate("/");
      return;
    }

    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    await registerLogin(provider, { access_token });
    reload();
    navigate("/");
  }, []);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return <h1>Please wait...</h1>;
}
