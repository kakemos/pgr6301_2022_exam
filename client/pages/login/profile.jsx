import React from "react";

export function Profile({ user }) {
  return (
    <>
      <h1>
        {user.google?.name
          ? `Profile for ${user.google.name}`
          : `Profile for ${user.microsoft.name}`}
      </h1>
      <h3>{user.google?.email ? user.google?.email : user.microsoft?.email}</h3>
      <div>
        <img src={user.google?.picture ? user.google?.picture : null} />
      </div>
    </>
  );
}
