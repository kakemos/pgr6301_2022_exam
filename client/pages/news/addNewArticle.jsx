import React, { useContext, useEffect, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useNavigate } from "react-router-dom";
import { FormInput, FormInputTextArea } from "../../lib/formInput";
import { useLoading } from "../../lib/useLoading";

export function AddNewArticle({ user }) {
  const { addNewArticle, listFilteredArticles } = useContext(NewsApiContext);
  const [title, setTitle] = useState("");
  const [intro, setIntro] = useState("");
  const [article, setArticle] = useState("");
  const [category, setCategory] = useState("general");
  const [author, setAuthor] = useState(user?.microsoft?.name);
  const [ws, setWs] = useState("");
  const navigate = useNavigate();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ title: title }),
    []
  );

  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    setWs(ws);
  }, []);

  async function handleSubmit(event) {
    event.preventDefault();
    if (title === "" || intro === "" || article === "") {
      return navigate("/error-empty-fields");
    } else if (title === data.title) {
      return navigate("error-duplicate-title");
    } else {
      await addNewArticle({ title, intro, article, category, author });
      ws.send(JSON.stringify({ title, intro }));
      navigate(`/article/${title}`);
    }
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  if (!user.microsoft) {
    return <Navigate to="/login" replace />;
  }
  return (
    <>
      <h1>Write new article.</h1>
      <form onSubmit={handleSubmit}>
        <h4>Author: {author}</h4>
        <FormInput label={"Title"} value={title} onChangeValue={setTitle} />
        <FormInput label={"Intro"} value={intro} onChangeValue={setIntro} />
        <div className={"form-input"}>
          <div>
            <label htmlFor="dog-names">
              <strong>Category</strong>
            </label>
          </div>
          <select
            className={"new-article-dropdown"}
            name="categories"
            id="category-dropdown"
            value={category}
            onChange={(e) => setCategory(e.target.value)}
          >
            <option value="general">General</option>
            <option value="fashion">Fashion</option>
            <option value="art">Art</option>
            <option value="travel">Travel</option>
            <option value="celebrity">Celebrity</option>
          </select>
        </div>
        <FormInputTextArea
          label={"Article"}
          value={article}
          onChangeValue={setArticle}
        />
        <div>
          <button className={"submit-article-button"}>Submit article</button>
        </div>
      </form>
    </>
  );
}
