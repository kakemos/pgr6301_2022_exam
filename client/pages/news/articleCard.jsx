import React from "react";

export function ArticleCard({
  news: { title, intro, author, category, article },
}) {
  return (
    <>
      <div className={"article-card"}>
        <h1 className={"article-title"}>{title}</h1>
        <hr />
        <div className={"article-under-header"}>
          <h4>Written by {author}</h4>
          <h4>category: {category}</h4>
        </div>
        <p>{intro}</p>
        <p>{article}</p>
      </div>
    </>
  );
}
