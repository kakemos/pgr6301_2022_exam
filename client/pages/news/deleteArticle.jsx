import React, { useContext, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useNavigate } from "react-router-dom";
import { useLoading } from "../../lib/useLoading";

export function DeleteArticle({ user }) {
  const { deleteArticle, listFilteredArticles } = useContext(NewsApiContext);
  const { title, setTitle } = useState("");
  const navigate = useNavigate();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ author: user?.microsoft?.name }),
    []
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  async function handleSubmit(event) {
    event.preventDefault();
    await deleteArticle({ title });
    navigate("/");
  }

  if (!user.microsoft) {
    return <Navigate to="/login" replace />;
  }

  return (
    <>
      <h2>Overview of all your articles.</h2>
      <h3>Which one do you wanna delete?</h3>
      <form onSubmit={handleSubmit}>
        {data.map((article) => (
          <>
            <input
              type={"radio"}
              name={"delete-articles"}
              value={article.title}
              onChange={() => setTitle(article.title)}
            />
            <label htmlFor={article.title}>{article.title}</label>
            <p>{article.intro}</p>
          </>
        ))}
        <button>Delete</button>
      </form>
    </>
  );
}
