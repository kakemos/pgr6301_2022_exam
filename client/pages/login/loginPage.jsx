import { Route, Routes, useNavigate } from "react-router-dom";
import React, { useContext, useEffect } from "react";
import { LoginButton } from "./loginButton";
import { LoginCallback } from "./loginCallback";
import { NewsApiContext } from "../../newsApiContext";

export function EndSession({ reload }) {
  const navigate = useNavigate();
  const { endSession } = useContext(NewsApiContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });
  return <h1>Please wait...</h1>;
}

function StartLogin({ config }) {
  return (
    <div>
      <h1>Login</h1>
      <LoginButton
        label={"Login with Google"}
        config={config}
        provider={"google"}
      />
      <LoginButton
        label={"Login with Microsoft"}
        config={config}
        provider={"microsoft"}
      />
    </div>
  );
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route path={"/"} element={<StartLogin config={config} />} />
      <Route
        path={"/:provider/callback"}
        element={<LoginCallback config={config} reload={reload} />}
      />
      <Route path={"/end-session"} element={<EndSession reload={reload} />} />
      <Route path={"*"} element={<StartLogin config={config} />} />
    </Routes>
  );
}
