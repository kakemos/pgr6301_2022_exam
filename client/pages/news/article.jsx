import React, { useContext } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useParams } from "react-router-dom";
import { useLoading } from "../../lib/useLoading";
import { ArticleCard } from "./articleCard";

export function Article({ user }) {
  const { listFilteredArticles } = useContext(NewsApiContext);
  const { title } = useParams();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ title: title }),
    []
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  if (!user.microsoft && !user.google) {
    return <Navigate to="/login" replace />;
  }

  return (
    <>
      {data.map((news) => (
        <>
          <ArticleCard key={news.title} news={news} />
        </>
      ))}
    </>
  );
}
