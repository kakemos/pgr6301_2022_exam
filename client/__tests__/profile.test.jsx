import ReactDOM from "react-dom";
import {MemoryRouter} from "react-router-dom";
import {AddNewArticle} from "../pages/news/addNewArticle";
import React from "react";
import {Profile} from "../pages/login/profile";

describe("profile component", () => {
  it("shows microsoft user profile", () => {
      const element = document.createElement("div");
      const user = { microsoft: { name: "Testy Testsen", email: "testy@test.com" } };

    ReactDOM.render(
        <MemoryRouter>
          <Profile user={user} />
        </MemoryRouter>,
        element
    );

    expect(element.querySelector("h1").innerHTML).toEqual("Profile for Testy Testsen");
    expect(element.querySelector("h3").innerHTML).toEqual("testy@test.com");
    expect(element.innerHTML).toMatchSnapshot();

  });

  it("shows google user profile", () => {
          const element = document.createElement("div");
          const user = { google: { name: "Testy Testsen", email: "testy@test.com" } };

          ReactDOM.render(
              <MemoryRouter>
                  <Profile user={user} />
              </MemoryRouter>,
              element
          );

          expect(element.querySelector("h1").innerHTML).toEqual("Profile for Testy Testsen");
          expect(element.querySelector("h3").innerHTML).toEqual("testy@test.com");
          expect(element.innerHTML).toMatchSnapshot();
  });
});
