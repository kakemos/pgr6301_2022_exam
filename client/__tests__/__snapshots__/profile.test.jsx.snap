// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile component shows google user profile 1`] = `"<h1>Profile for Testy Testsen</h1><h3>testy@test.com</h3><div><img></div>"`;

exports[`profile component shows microsoft user profile 1`] = `"<h1>Profile for Testy Testsen</h1><h3>testy@test.com</h3><div><img></div>"`;
